src/io/mod.rs

```diff
@@ -88,15 +92,18 @@ impl Io {
 
     /// Opens a file for reading.  Transparently handles reading gzipped files based
     /// extension.
-    pub fn new_reader<P>(&self, p: &P) -> Result<BufReader<Box<dyn Read>>>
+    pub fn new_reader<P>(&self, p: &P) -> Result<Box<dyn BufRead>>
     where
         P: AsRef<Path>,
     {
         let file = File::open(p).map_err(FgError::IoError)?;
-        let read: Box<dyn Read> =
-            if Io::is_gzip_path(p) { Box::new(MultiGzDecoder::new(file)) } else { Box::new(file) };
+        let buf = BufReader::with_capacity(self.buffer_size, file);
 
-        Ok(BufReader::new(read))
+        if Self::is_gzip_path(p) {
+            Ok(Box::new(BufReader::with_capacity(self.buffer_size, MultiGzDecoder::new(buf))))
```

Comment:
This causes us to buffer twice, but at least performance wise, it's a non-trivial improvement over what was there before.  So we have one buffer for reading from the File (100% necessary) and another buffer wrapped around the MultiGzDecoder because it only implements `Read` and not `BufRead`.